### PR TITLE
misc/babysuprem.cpp: Now working machine

### PR DESCRIPTION
**Pull request by Paul-Arnold**
Update driver to make Baby Suprem a playable game.
There is a basic layout file which allows the game to be played but a proper layout with artwork is really needed in the future.
Not sure I've used the correct flags for this where the game is declared ?

### CODE REVIEW

File: src/mame/misc/babysuprem.cpp [hunk 1]
```diff
@@ -14,11 +14,24 @@
...
+		  There are a six lamps whose use needs identifying.
+		  Identify what's being accessed in the i/o map, this doesn't appear to
+		  affect the game at all.
+		  The manual lists all chips used and includes a Z80 DART and Z80 CTC neither
+		  of which appear in the schematic.
 */
 
 #include "emu.h"
-
+#include "babysuprem.lh"
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:31]: Please put “headers” that actually contain definitions after all the declaration headers.  Also, the blank lines separating grouped headers are intentional.

File: src/mame/misc/babysuprem.cpp [hunk 2]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
+	uint8_t ay8910_w_dsw1_switches2_r(offs_t offset);
+	void ay8910_w(offs_t offset, uint8_t data);
+
+	uint8_t dsw2_u13_switches_r(offs_t offset);
+	void u9_w(offs_t offset, uint8_t data);
+	void u10_w(offs_t offset, uint8_t data);
+
+	virtual void machine_reset() override;
+	virtual void machine_start() override;
+
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:107]: Please keep overridden virtual member functions at the same visibility level.  These should be protected.

File: src/mame/misc/babysuprem.cpp [hunk 3]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
 static INPUT_PORTS_START( bsuprem )
-	PORT_START("DSW1")
-	PORT_DIPUNKNOWN_DIPLOC(0x01, 0x01, "SW1:1")
-	PORT_DIPUNKNOWN_DIPLOC(0x02, 0x02, "SW1:2")
-	PORT_DIPUNKNOWN_DIPLOC(0x04, 0x04, "SW1:3")
-	PORT_DIPUNKNOWN_DIPLOC(0x08, 0x08, "SW1:4")
+	PORT_START("DSW1_SWITCHES2")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_BUTTON3) PORT_NAME("Right stop")
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_BUTTON2) PORT_NAME("Middle stop")
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_BUTTON1) PORT_NAME("Left stop")
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:116]: Please use the proper input types for slot machine controls.

File: src/mame/misc/babysuprem.cpp [hunk 4]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
+	PORT_DIPSETTING(   0x80, DEF_STR( Off ))
+	PORT_DIPNAME(0x40, 0x40, "SW1-1 (Test 2)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x40, DEF_STR( Off ))
+	PORT_DIPNAME(0x20, 0x20, "SW1-2 (Test 3)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x20, DEF_STR( Off ))
+	PORT_DIPNAME(0x10, 0x10, "SW1-4 (Test 5)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x10, DEF_STR( Off ))
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:129]: If these are DIP switches, please use DIP location markup properly so switches will be shown on the menu.

File: src/mame/misc/babysuprem.cpp [hunk 5]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
+	PORT_DIPSETTING(   0x00, DEF_STR( Yes ))
+	PORT_DIPSETTING(   0x02, DEF_STR( No ))
+	PORT_DIPNAME(0x01, 0x01, "SW2-1 (Bet/Collect)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x01, DEF_STR( Off ))
+
+	PORT_START("U13_SWITCHES")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_BUTTON5)  PORT_NAME("Bet/Collect")
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_BUTTON6)  PORT_NAME("Start")
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_SERVICE)  PORT_NAME("Door") PORT_CODE(KEYCODE_D) PORT_TOGGLE
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:160]: Same here, please use appropriate input types, and use proper markup for DIP switches.

File: src/mame/misc/babysuprem.cpp [hunk 6]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
+
+void bsuprem_state::machine_start()
+{
+	m_digits.resolve();
+	m_lamps.resolve();
+	m_buttons.resolve();
+	m_nixie.resolve();
+
+	m_nvram8 = std::make_unique<uint8_t[]>(2048);
+	m_nvram->set_base(m_nvram8.get(),2048);
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:233]: Do you need to register this for save states as well?  Also, please put a space after the comma.

File: src/mame/misc/babysuprem.cpp [hunk 7]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
+	// bit 6 of u10 is 1. However, implementing this makes it impossible
+	// to enter test/setup. As there are a number of other errors on
+	// the schematic it's likely this is wrong.
+	m_nvram8[offset] = data;
+}
+
+uint8_t bsuprem_state::nvram_r(offs_t offset)
+{
+	return m_nvram8[offset];
+}
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:252]: Do you actually need this trampoline?  Could you install it as ram with `.share(...)` so you can get a convenient pointer to it, and then not need to allocate/save it yourself either?

File: src/mame/misc/babysuprem.cpp [hunk 8]
```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
...
+	// bit 6 = hopper 5
+	// bit 7 = hopper 25
+	m_buttons[0]=BIT(data,4);
+	m_buttons[1]=BIT(data,2);
+	m_buttons[2]=BIT(data,1);
+	m_buttons[3]=BIT(data,0);
+	m_buttons[4]=BIT(data,4);
+
+	m_hopper_5->motor_w(BIT(data, 6));
+	m_hopper_25->motor_w(BIT(data, 7));
```
Comment (cuavas) [inline src/mame/misc/babysuprem.cpp:363]: Please use spaces around binary operators and after commas.

File: src/mame/layout/babysuprem.lay [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+		<element name="digit5" ref="digit">
+			<bounds x="150" y="15" width="4" height="10"/>
+		</element>
+
+<!--
+left hand bonus coins
+-->
+	<element name="lamp30" ref="coinlp">
+			<bounds x="20" y="70" width="7" height="7"/>
+		</element>
```
[the remark this answers is not in the document]
Comment (Paul-Arnold) [reply src/mame/layout/babysuprem.lay:301]: Yes, originally everything was in numerical order and I found it easier when moving items into groups to have each group easy to see.
I'll fix the indentation and ident the comment though.

File: src/mame/misc/babysuprem.cpp (hunk 1 above)
Comment (Paul-Arnold) [reply src/mame/misc/babysuprem.cpp:31]: OK, didn't realise about the blank lines, some drivers have them and others don't. Will fix.

File: src/mame/misc/babysuprem.cpp (hunk 2 above)
Comment (Paul-Arnold) [reply src/mame/misc/babysuprem.cpp:107]: OK, I was unlucky to pick a driver for reference which had them  in the wrong place :(

**Pull request merged**